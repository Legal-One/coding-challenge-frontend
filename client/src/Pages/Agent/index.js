import React, { useState, useEffect } from "react";
import axios from "axios";
import { useParams } from "react-router-dom";
import AgentCallLogs from "../../components/AgentCallLogs";
import "./index.scss";


const apiUrl = "http://localhost:5000/api/agent/";

function Agent() {
    const { agentId } = useParams();
    
    const [agentCallLogs, setAgentCallLogs] = useState([]);

    useEffect(() => {
        const fetchAgentCallLogs = async () => {
            const response = await axios.get(apiUrl + agentId);
            setAgentCallLogs(response.data);
        };
        fetchAgentCallLogs();
    }, [agentId]);



    return (
        <div>
            <AgentCallLogs agentCallLogs = {agentCallLogs} />
        </div>
    );
}
export default Agent;


