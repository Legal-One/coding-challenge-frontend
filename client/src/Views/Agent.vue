<template>
  <div class="agent-content">
    <h2 class="">Agent Details</h2>
    <div v-if="isLoading">
      Loaging
    </div>
    <AgentDetails v-if="!isLoading" :agent="agent"></AgentDetails>
    <h2 class="">Agent Logs</h2>
    <div class="agent-content__logs" v-if="!isLoading">
      <AgentLogItem v-for="(log, index) in logs" :phoneLog="log" :key="index"></AgentLogItem>
    </div>
  </div>
</template>

<script>
import AgentDetails from '@/components/AgentDetails.vue';
import AgentLogItem from '@/components/AgentLogItem.vue';

export default {
  name: 'Agent',
  data() {
    return {
      agentDetails: {},
      agentLogs: [],
      loading: false,
    };
  },
  computed: {
    isLoading: {
      get() {
        return this.loading;
      },
      set(value) {
        this.loading = value;
      },
    },
    agent: {
      get() {
        return this.agentDetails;
      },
      set(value) {
        this.agentDetails = value;
      },
    },
    logs: {
      get() {
        return this.agentLogs;
      },
      set(value) {
        this.agentLogs = value;
      },
    },
  },
  methods: {
    loadAgentDetails() {
      const self = this;
      console.log('env ', process.env.VUE_APP_API_URL);
      const { id } = this.$route.params;
      fetch(`${process.env.VUE_APP_API_URL}agent/${id}`)
        .then((response) => response.json())
        .then(({ status, agent, logs }) => {
          if (status === 200) {
            self.agent = agent;
            self.logs = logs;
          }
        });
    },
  },
  created() {
    this.loadAgentDetails();
  },
  components: {
    AgentDetails,
    AgentLogItem,
  },
};
</script>

<style lang="scss">
  .agent-content {
    &__logs {
      margin-top: 35px;
    }
  }
</style>
