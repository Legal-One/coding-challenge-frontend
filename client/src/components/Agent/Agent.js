import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { getAgentLogs } from "../../api";

const Agent = () => {
  const [logs, setLogs] = useState([]);
  let { id } = useParams();

  useEffect(() => {
    const fetchLogs = async () => {
      const logs = await getAgentLogs(id);
      console.log(logs)
      if (logs.length) {
        return setLogs(logs);
      }
    };
    fetchLogs();
  }, [id]);

  return (
    <div className="agent">
      <table>
        <thead>
          <tr>
            <th>Phone number</th>
            <th>Call date and time</th>
            <th>Resolution</th>
          </tr>
        </thead>
        <tbody>
          {logs.map((log) => (
            <tr key={log.identifier}>
              <th>{log.number}</th>
              <th>{log.dateTime}</th>
              <th>{log.resolution.resolution}</th>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Agent;
