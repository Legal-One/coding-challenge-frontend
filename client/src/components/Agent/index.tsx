import React from 'react'
import { useParams } from 'react-router-dom'
import AgentStats from './agent-stats'

function Agent() {
  const { agentId } = useParams<Record<string, string>>()

  return (
    <div>
      <AgentStats agentId={agentId} />
    </div>
  )
}

export default Agent
