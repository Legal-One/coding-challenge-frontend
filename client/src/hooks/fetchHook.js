import { useState, useEffect } from "react";
import axios from "axios";

export const useFetchData = (apiUrl) => {
    const [data, setData] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);

    useEffect(() => {
        const fetchData = async () => {
            setIsError(false);
            setIsLoading(true);

            try {
                const result = await axios(apiUrl);
                setData(result.data);
            } catch (error) {
                setIsError(true);
            }

            setIsLoading(false);
        };

        fetchData();
    }, [apiUrl]);

    return [data, isLoading, isError];
};
