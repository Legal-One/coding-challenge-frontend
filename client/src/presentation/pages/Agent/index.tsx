import { FC } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { ArrowIcon } from '../../../assets/SVG';

import UserInfo from './components/UserInfo';
import CallData from './components/CallData';
import useFetchAgentData from './hooks/useFetchAgentData';

const Agent: FC = () => {
	const params: any = useParams();
	const { agentData, agentLogs } = useFetchAgentData(params.id);
	const { goBack } = useHistory();
	const onGoBackClick = () => {
		goBack();
	};

	return (
		<main className="callPage page__wrapper">
			<div className="page__container">
				<section className="callPage__header">
					<button className="goBack" onClick={onGoBackClick}>
						<ArrowIcon />
						<p>Go Back</p>
					</button>
					<h2>
						Agent: <span>{agentData ? `${agentData.firstName} ${agentData.lastName}` : ''}</span>
					</h2>
				</section>
				<UserInfo data={agentData} logs={agentLogs} />
				<CallData records={agentLogs} />
			</div>
		</main>
	);
};

export default Agent;
