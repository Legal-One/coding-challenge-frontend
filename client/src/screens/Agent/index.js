import React from 'react'
import MapComponent from '../../components/MapComponent';
import Table from '../../components/Table';
import { agentCallLogs } from '../../utils';
import { useHistory, useLocation } from 'react-router-dom';
import { MAP_AXIS, MAP_TITLE, TABLE_HEADINGS } from '../../utils/constants';

function Agent({ logs, agents, resolution }) {
    const history = useHistory();
    const location = useLocation();
    const agentId = location.pathname.split('/')[2];
    const calculatedData = agentCallLogs({ calls: logs.filter(log => log.agentIdentifier === agentId), agent: agents[agentId], resolution });

    return (
        <div>
            <MapComponent data={calculatedData.mapData} axis={MAP_AXIS.AGENT} title={MAP_TITLE.AGENT}/>
            <Table headings={TABLE_HEADINGS.AGENT} >
                {
                    calculatedData.tableData.map((call, index) => (
                        <tr key={call.identifier}>
                            <td><span onClick={() => history.push(`/call/${call.number}`)}>
                                {call.number}
                            </span></td>
                            <td>{call.date} {call.time}</td>
                            <td>{call.resolution}</td>
                        </tr>
                    ))
                }
            </Table>
        </div>
    )
}

export default Agent;
