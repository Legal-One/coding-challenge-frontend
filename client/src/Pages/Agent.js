import React from "react";
import { useParams } from "react-router-dom";
import { useFetchData } from "../hooks/useFetchData";
import Table from "../components/Table";
import Chart from "../components/Chart";
import { getFormattedDateTime } from "../utils/timeFormat";
import "./index.scss";

const apiUrl = "/api/agent/";
const headers = ["Phone number", "Call date and time", "Resolution"];
const dataMapper = (agentData) => {
    return agentData.map((agent) => {
        return [
            agent.phoneNumber,
            getFormattedDateTime(agent.callTime),
            agent.resolution,
        ];
    });
};

function Agent() {
    const { agentId, agentName } = useParams();
    const title = `Call history for ${agentName}`;
    const [data, isLoading, isError] = useFetchData(apiUrl + agentId);

    const counter = data.reduce((acc, cur) => {
        if (acc[cur.resolution]) {
            acc[cur.resolution]++;
        } else {
            acc[cur.resolution] = 1;
        }
        return acc;
    }, {});

    const chartData = [["Resolution", "Count"], ...Object.entries(counter)];

    return (
        <main>
            {!isLoading && !isError && (
                <>
                    <Table
                        title={title}
                        headers={headers}
                        rows={dataMapper(data)}
                    />
                    <Chart chartType="PieChart" chartData={chartData} />
                </>
            )}
            {isLoading && <h3>Fetching data...</h3>}
            {isError && <h3>Error fetching data. Please refresh your page</h3>}
        </main>
    );
}
export default Agent;
