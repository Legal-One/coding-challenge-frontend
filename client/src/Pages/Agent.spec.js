import React from "react";
import { render } from "@testing-library/react";
import Agent from "./Agent";
import * as mockHook from "../hooks/fetchHook";

jest.mock("react-router-dom", () => ({
    useParams: () => ({
        agentId: "beb04834-9f8b-11e7-9469-8a39b455b609",
        agentName: "John Doe",
    }),
}));

describe("AgentPage", () => {
    test("render with all props passed", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [
                {
                    phoneNumber: "+4912345678",
                    callTime: "2020-10-06T13:50:00.000Z",
                    resolution: "mock-resolution",
                },
                {
                    phoneNumber: "+49987654321",
                    callTime: "2020-11-06T13:50:00.000Z",
                    resolution: "mock-resolution-2",
                },
            ],
            false,
            false,
        ]);
        const { asFragment } = render(<Agent />);
        expect(asFragment()).toMatchSnapshot();
    });

    test("render with error from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            false,
            true,
        ]);
        const { asFragment } = render(<Agent />);
        expect(asFragment()).toMatchSnapshot();
    });

    test("render when waiting for response from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            true,
            false,
        ]);
        const { asFragment } = render(<Agent />);
        expect(asFragment()).toMatchSnapshot();
    });
});
