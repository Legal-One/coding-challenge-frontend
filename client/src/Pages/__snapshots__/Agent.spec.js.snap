// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentPage render when waiting for response from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Fetching data...
    </h3>
  </main>
</DocumentFragment>
`;

exports[`AgentPage render with all props passed 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Call history for John Doe
    </h3>
    <table
      class="Table"
    >
      <thead>
        <tr>
          <th>
            Phone number
          </th>
          <th>
            Call date and time
          </th>
          <th>
            Resolution
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            +4912345678
          </td>
          <td>
            10/6/2020 3:50:00 PM
          </td>
          <td>
            mock-resolution
          </td>
        </tr>
        <tr>
          <td>
            +49987654321
          </td>
          <td>
            11/6/2020 2:50:00 PM
          </td>
          <td>
            mock-resolution-2
          </td>
        </tr>
      </tbody>
    </table>
  </main>
</DocumentFragment>
`;

exports[`AgentPage render with error from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Error fetching data. Please refresh your page
    </h3>
  </main>
</DocumentFragment>
`;
